File: ios/Classes/SecureKeyPlugin.swift
import Flutter
import UIKit
import AudioToolbox

public class SecureKeyPlugin: NSObject, FlutterPlugin {
    var controller:AppKeyPair?
    
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "secure_key", binaryMessenger: registrar.messenger())
    let instance = SecureKeyPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
     
      if call.method == "initialize"{
          initialize(call:call,result: result)
      }else if call.method == "getPublicKey"{
          getPublicKey(result: result)
      }else if call.method == "createPairKey"{
          createPairKey(result: result)
      }else if call.method == "deleteKey"{
          deleteKey(result: result)
      }else if call.method == "signSha256"{
          signSha256(call: call, result: result)
      }else if call.method == "getPublicKeyBytes"{
          getPublicKeyBytes(result: result)
      }else if call.method == "signSha256Bytes"{
          signSha256Bytes(call: call, result: result)
      }else if call.method == "encryptWithRsa"{
          encryptWithRsa(call: call, result: result)
      }else if call.method == "decryptWithRsa" {
          decryptWithRsa(call: call, result: result)
      } else{
          result(FlutterMethodNotImplemented)
      }
  }
    
    private func initialize(call:FlutterMethodCall,result: @escaping FlutterResult){
        if controller != nil {
          result(true)
          return;
        }
        if let args = call.arguments as? Dictionary<String, Any>,
           let size = args["size"] as? Int
        {
            controller = AppKeyPair(size: size)
            result(true)
        }else{
            result(FlutterError(code: String(describing: KeyPairError.BAD_ARGS),message: "Size argument not found",details: nil))
        }
    }
    
    private func getPublicKey(result:@escaping FlutterResult){
        guard controller != nil else{
            result(FlutterError(code: String(describing: KeyPairError.NOT_INIT), message: "Plugin not init", details: nil))
          return
        }
        do{
            let publicKey:String? = try controller!.getPublicKey()
            
            guard !(publicKey ?? "").isEmpty else{
                result(FlutterError(code: String(describing: KeyPairError.NOT_FOUND) ,message: "Public key not found",details: nil))
                return
            }
            result(publicKey)
        }catch
            KeyPairError.NOT_FOUND{
            result(FlutterError(code: "PUBLIC_NOT_FOUND", message: "Public key not founded", details: nil))
        }
        catch{
            result(FlutterError(code: "UNKNOWN", message: "Unknown error", details: nil))
        }
       }
    
    private func getPublicKeyBytes(result:@escaping FlutterResult){
        guard controller != nil else{
            result(FlutterError(code: String(describing: KeyPairError.NOT_INIT), message: "Plugin not init", details: nil))
          return
        }
        do{
            let publicKey:FlutterStandardTypedData = try controller!.getPublicKeyBytes()
            result(publicKey)
        }
//        catch
//            KeyPairError.NOT_FOUND{
//            result(FlutterError(code: "PUBLIC_NOT_FOUND", message: "Public key not founded", details: nil))
//        }
        catch{
            result(nil)
//            result(FlutterError(code: "UNKNOWN", message: "Unknown error", details: nil))
        }
        }

    private func createPairKey(result:@escaping FlutterResult){
        guard controller != nil else{
            result(FlutterError(code: String(describing: KeyPairError.NOT_INIT), message: "Plugin not init", details: nil))
          return
        }
        do{
            let created:Bool = try controller!.createRsaKeyPair()
            result(created)
        }catch KeyPairError.CREATE_FAIL{
            result(FlutterError(code: String(describing: KeyPairError.CREATE_FAIL), message: "Key pair created fail", details: nil))
        }
        catch{
            result(FlutterError(code: String(describing: KeyPairError.UNKNOWN), message: "Unknown error", details: nil))
        }
      }
    
    private func deleteKey(result:@escaping FlutterResult) {
        guard controller != nil else{
            result(FlutterError(code: String(describing: KeyPairError.NOT_INIT), message: "Plugin not init", details: nil))
          return
        }
        do{
          let deleted:Bool = try controller!.deleteKey()
          result(deleted)
        }catch KeyPairError.REMOVE_FAIL{
            result(FlutterError(code: String(describing: KeyPairError.REMOVE_FAIL), message: "Delete key fail", details: nil))
        }catch{
            result(FlutterError(code: String(describing: KeyPairError.UNKNOWN), message: "Unknown error", details: nil))
        }
    }
    
    private func signSha256(call:FlutterMethodCall,result: @escaping FlutterResult){
        guard controller != nil else{
            result(FlutterError(code: String(describing: KeyPairError.NOT_INIT), message: "Plugin not init", details: nil))
          return
        }
        if let args = call.arguments as? Dictionary<String, Any>,
           let input = args["inputSha256"] as? String
        {
            do{
                let signature:String? = try controller!.signSha256(input: input)
                guard signature != nil else{
                    result(FlutterError(code: String(describing: KeyPairError.SIGNATURE_FAIL), message: "Signature not founded", details:nil))
                    return
                }
                result(signature)
            }catch KeyPairError.SIGNATURE_FAIL{
                result(FlutterError(code: String(describing: KeyPairError.SIGNATURE_FAIL), message: "Signature not founded", details:nil))
            }catch{
                result(FlutterError(code: String(describing: KeyPairError.UNKNOWN), message: "Unknown error", details: nil))
            }
           
        }else{
            result(FlutterError(code: String(describing: KeyPairError.BAD_ARGS),message: "Input argument not found",details: nil))
        }
    }
    
    private func encryptWithRsa(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard controller != nil else {
            result(FlutterError(code: String(describing: KeyPairError.NOT_INIT), message: "Plugin not init", details: nil))
          return
        }
        if let args = call.arguments as? Dictionary<String, Any>,
           let input = args["encryptInput"] as? String
        {
            do{
                let encryptedData:String? = try controller!.encryptWithRsa(plainText: input)
                result(encryptedData)
            }catch KeyPairError.ALGORITM_NOT_SUPPORTED {
                result(FlutterError(code: String(describing: KeyPairError.ALGORITM_NOT_SUPPORTED), message: "Algoritm support error", details: nil))
            }catch KeyPairError.NOT_FOUND{
                result(FlutterError(code: String(describing: KeyPairError.NOT_FOUND), message: "Public key not found", details: nil))
            }catch KeyPairError.ENCRYPT_ERROR {
                result(FlutterError(code: String(describing: KeyPairError.ENCRYPT_ERROR), message: "Encrypt failed", details: nil))
            }catch KeyPairError.INPUT_NOT_FOUND {
                result(FlutterError(code: String(describing: KeyPairError.INPUT_NOT_FOUND), message: "Encrypt failed, input nil", details: nil))
            }catch {
                result(FlutterError(code: String(describing: KeyPairError.ENCRYPT_ERROR), message: "Encryption failed, Error occurred: \(error)", details: nil))
            }
           
        }else{
            result(FlutterError(code: String(describing: KeyPairError.BAD_ARGS),message: "Input argument not found",details: nil))
        }
    }
    
    private func decryptWithRsa(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard controller != nil else {
            result(FlutterError(code: String(describing: KeyPairError.NOT_INIT), message: "Plugin not init", details: nil))
          return
        }
        if let args = call.arguments as? Dictionary<String, Any>,
           let input = args["decryptInput"] as? String
        {
            do{
                let decryptedData:String? = try controller!.decryptWithRsa(cipherData: input)
                result(decryptedData)
            }catch KeyPairError.ALGORITM_NOT_SUPPORTED {
                result(FlutterError(code: String(describing: KeyPairError.ALGORITM_NOT_SUPPORTED), message: "Algoritm support error", details: nil))
            }catch KeyPairError.NOT_FOUND{
                result(FlutterError(code: String(describing: KeyPairError.NOT_FOUND), message: "Private key not found", details: nil))
            }catch KeyPairError.DECRYPT_ERROR {
                result(FlutterError(code: String(describing: KeyPairError.DECRYPT_ERROR), message: "Decrypt failed", details: nil))
            }catch {
                result(FlutterError(code: String(describing: KeyPairError.DECRYPT_ERROR), message: "Encryption failed, Error occurred: \(error)", details: nil))
            }
           
        }else{
            result(FlutterError(code: String(describing: KeyPairError.BAD_ARGS),message: "Input argument not found",details: nil))
        }
    }
    
    
    
    private func signSha256Bytes(call:FlutterMethodCall,result: @escaping FlutterResult){
        guard controller != nil else{
            result(FlutterError(code: String(describing: KeyPairError.NOT_INIT), message: "Plugin not init", details: nil))
          return
        }
        if let args = call.arguments as? Dictionary<String, Any>,
           let input = args["inputSha256"] as? String
        {
            do{
                let signature:FlutterStandardTypedData? = try controller!.signSha256Bytes(input: input)
                result(signature)
            }catch KeyPairError.SIGNATURE_FAIL{
                result(FlutterError(code: String(describing: KeyPairError.SIGNATURE_FAIL), message: "Signature not founded", details:nil))
            }catch{
                result(FlutterError(code: String(describing: KeyPairError.UNKNOWN), message: "Unknown error", details: nil))
            }
           
        }else{
            result(FlutterError(code: String(describing: KeyPairError.BAD_ARGS),message: "Input argument not found",details: nil))
        }
    }
    
}

File: android/src/main/java/kg/storekey/secure_key/SecureKeyPlugin.java
package kg.storekey.secure_key;

import android.content.Context;
import androidx.annotation.NonNull;

import java.security.InvalidKeyException;
import java.security.NoSuchAlgorithmException;
import java.security.NoSuchProviderException;

import javax.crypto.BadPaddingException;
import javax.crypto.IllegalBlockSizeException;
import javax.crypto.NoSuchPaddingException;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.Result;


/** SecureKeyPlugin */
public class SecureKeyPlugin implements FlutterPlugin, MethodChannel.MethodCallHandler {

  private MethodChannel channel;
  private Context context;
  private AppKeyPair controller;


  @Override
  public void onAttachedToEngine(@NonNull FlutterPluginBinding flutterPluginBinding) {
    channel = new MethodChannel(flutterPluginBinding.getBinaryMessenger(), "secure_key");
    context =  flutterPluginBinding.getApplicationContext();
    channel.setMethodCallHandler(this);
  }


  @Override
  public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {
    switch (call.method){
      case "initialize":
        initialize(call, result);
        break;
      case "getPublicKey":
        getPublicKey(result);
        break;
      case "createPairKey":
        createKeyPair(result);
        break;
      case "deleteKey":
        deleteKey(result);
        break;
      case "signSha256":
        signSha256(call,result);
        break;
      case "getPublicKeyBytes":
        getPublicKeyBytes(result);
        break;
      case "signSha256Bytes":
        signSha256Bytes(call,result);
        break;
      case "encryptWithRsa":
        encryptWithRsa(call, result);
        break;
      case "decryptWithRsa":
        decryptWithRsa(call, result);
        break;
      default:
        result.notImplemented();
    }
  }

  private void initialize(MethodCall call,Result result){
    if(controller != null){
      result.success(true);
      return;
    }
    Integer size = call.argument("size");
    if(size != null){
      controller = new AppKeyPair(context, size);
      result.success(true);

    }else {
      result.error(AppKeyPairErrors.BAD_ARGS.toString(), "Size argument mush be NonNull", null);
    }
  }


  private void deleteKey(Result result){
    if(controller == null){
      result.error(AppKeyPairErrors.NOT_INIT.toString(), "Controller not init", null);
      return;
    }
    try{
      boolean deleted = controller.deleteKeyPair();
      result.success(deleted);
    }catch (RuntimeException e){
      result.error(AppKeyPairErrors.REMOVE_FAIL.toString(),"Key pair removed fail",null);
    }

  }

  private void getPublicKey(Result result){
    if(controller == null){
      result.error(AppKeyPairErrors.NOT_INIT.toString(), "Controller not init", null);
      return;
    }
    String key = controller.getPublicKey();
    result.success(key);
  }

  private void getPublicKeyBytes(Result result){
    if(controller == null){
      result.error(AppKeyPairErrors.NOT_INIT.toString(), "Controller not init", null);
      return;
    }
    byte[] key = controller.getPublicKeyBytes();
    result.success(key);
  }

  private void encryptWithRsa(MethodCall call, Result result) {
    if(controller == null) {
      result.error(AppKeyPairErrors.NOT_INIT.toString(), "Controller not init", null);
      return;
    }
    String input = call.argument("encryptInput");
    if(input != null) {
      try {
        String encryptedData = controller.encryptWithRsa(input);
        if(encryptedData != null) {
          result.success(encryptedData);
        }else  {
          result.error(AppKeyPairErrors.ENCRYPTION_FAIL.toString(),"DATA NOT ENCRYPTED!",null);
        }
      } catch (NoSuchPaddingException | IllegalBlockSizeException | NoSuchAlgorithmException |
               BadPaddingException | InvalidKeyException |NoSuchProviderException e) {
        result.error(AppKeyPairErrors.ENCRYPTION_FAIL.toString(),"DATA NOT ENCRYPTED " + e.toString(),null);
      }
    }
  }

  private void decryptWithRsa(MethodCall call, Result result) {
    if(controller == null){
      result.error(AppKeyPairErrors.NOT_INIT.toString(), "Controller not init", null);
      return;
    }
    String input = call.argument("decryptInput");
    if(input != null) {
      try {
       String decryptedData =  controller.decryptWithRsa(input);
       if(decryptedData != null) {
         result.success(decryptedData);
       }else {
         result.error(AppKeyPairErrors.DECRYPTION_FAIL.toString(),"DATA NOT DECRYPTED!",null);
       }
      }catch (NoSuchAlgorithmException|
      NoSuchPaddingException|
              InvalidKeyException|
              IllegalBlockSizeException|
              BadPaddingException        |NoSuchProviderException  e) {

        result.error(AppKeyPairErrors.DECRYPTION_FAIL.toString(),"DATA NOT DECRYPTED " + e.toString(),null);
      }
    }
  }
  private void signSha256(MethodCall call,Result result){
    if(controller == null){
      result.error(AppKeyPairErrors.NOT_INIT.toString(), "Controller not init", null);
      return;
    }
    String input = call.argument("inputSha256");
    if(input != null){
      try{
        String sign = controller.signSha256(input);
        if(sign != null){
          result.success(sign);
        }else {
          result.error(AppKeyPairErrors.SIGNATURE_FAIL.toString(),"Signature not created",null);
        }
      }catch (RuntimeException e){
        if(e.getMessage() != null && e.getMessage().equals(AppKeyPairErrors.PRIVATE_KEY_NOT_FOUND.toString())){
          result.error(AppKeyPairErrors.PRIVATE_KEY_NOT_FOUND.toString(),"Private key not found in storage",null);
        }else{
          result.error(AppKeyPairErrors.SIGNATURE_FAIL.toString(),"Signature not created",null);
        }
      }
    }else{
      result.error(AppKeyPairErrors.BAD_ARGS.toString(),"Input data not found",null);
    }
  }

  private void signSha256Bytes(MethodCall call,Result result){
    if(controller == null){
      result.error(AppKeyPairErrors.NOT_INIT.toString(), "Controller not init", null);
      return;
    }
    String input = call.argument("inputSha256");
    if(input != null){
      try{
        byte[] sign = controller.signSha256Bytes(input);
        if(sign != null){
          result.success(sign);
        }else {
          result.error(AppKeyPairErrors.SIGNATURE_FAIL.toString(),"Signature not created",null);
        }
      }catch (RuntimeException e){
        if(e.getMessage() != null && e.getMessage().equals(AppKeyPairErrors.PRIVATE_KEY_NOT_FOUND.toString())){
          result.error(AppKeyPairErrors.PRIVATE_KEY_NOT_FOUND.toString(),"Private key not found in storage",null);
        }else{
          result.error(AppKeyPairErrors.SIGNATURE_FAIL.toString(),"Signature not created",null);
        }
      }
    }else{
      result.error(AppKeyPairErrors.BAD_ARGS.toString(),"Input data not found",null);
    }
  }


  private void createKeyPair(Result result){
    if(controller == null){
      result.error(AppKeyPairErrors.NOT_INIT.toString(), "Controller not init", null);
      return;
    }
    try {
      controller.createKeyPair();
      String key = controller.getPublicKey();
      if(key != null){
        result.success(true);
      } else {
        result.error(AppKeyPairErrors.CREATE_FAIL_NOT_FOUND.toString(),"Public key not found after created",null);
      }
    }catch (RuntimeException e){
      result.error(AppKeyPairErrors.CREATE_FAIL.toString(),"Fail when create key pair",null);
    }
  }




  @Override
  public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
    channel.setMethodCallHandler(null);
  }

}


